chore(docs): Migrating from Apollo Guide (#123)
nearform/graphql-hooks: README.md

}
```

## Authentication

Coming soon!

## Fragments

Coming soon!

## Migrating from Apollo

For a real life example, compare the next.js [with-apollo](https://github.com/zeit/next.js/tree/canary/examples/with-apollo) vs [with-graphql-hooks](https://github.com/zeit/next.js/tree/canary/examples/with-graphql-hooks). We have feature parity and the `main-*.js` bundle is a whopping **93% smaller** (7.9KB vs 116KB).

### ApolloClient ➡️ GraphQLClient

```diff
- import { ApolloClient } from 'apollo-client'
- import { InMemoryCache } from 'apollo-cache-inmemory'
+ import { GraphQLClient } from 'graphql-hooks'
+ import memCache from 'graphql-hooks-memcache'

- const client = new ApolloClient({
-  uri: '/graphql',
-  cache: new InMemoryCache()
- })
+ const client = new GraphQLClient({
+   url: '/graphql',
+   cache: memCache()
+ })
```

> Alot of the options you'd pass to `ApolloClient` are the same as `GraphQLClient`:

Reviewer: ```suggestion
A lot of the options you'd pass to `ApolloClient` are the same as `GraphQLClient`:
```

- `uri` ➡️ `url`
- `fetchOptions`
- `onError` - the function signature is slighly different
- `headers`
- `fetch`
- `cache`

### ApolloProvider ➡️ ClientContext.Provider

> ```diff

Reviewer: Didn't realise you could do this, it's awesome 👏

- import { ApolloProvider } from 'react-apollo'
+ import { ClientContext } from 'graphql-hooks'

function App({ client }) {
  return (
-    <ApolloProvider client={client}>
+    <ClientContext.Provider value={client}>
       {/* children */}
+    </ClientContext.Provider>
-    </ApolloProvider>
  )
}
```

### Query Component ➡️ useQuery

```diff
- import { Query } from 'react-apollo'
- import gql from 'graphql-tag'
+ import { useQuery } from 'graphql-hooks'

function MyComponent() {
+ const { loading, error, data } = useQuery('...')

-  return (
-    <Query query={gql`...`}>
-     {({ loading, error, data}) => {
        if (loading) return 'Loading...'
        if (error) return 'Error :('

        return <div>{data}</div>
-      }}
-    </Query>
-  )
}
```

### Query Component Props

> Alot of options can be carried over as-is, or have direct replacement:

Reviewer: ```suggestion
A lot of options can be carried over as-is, or have direct replacements:
```

- `query` ➡️ `useQuery(query)`: No need to wrap the query in `gql`
- `variables` ➡️ `useQuery(query, { variables })`
- `ssr` ➡️ `useQuery(query, { ssr })`

> - **Fetch Policies**: See [#75](https://github.com/nearform/graphql-hooks/issues/75) for a more info

Reviewer: ```suggestion
- **Fetch Policies**: See [#75](https://github.com/nearform/graphql-hooks/issues/75) for more info
```

  - `cache-first`: This the default behaviour of `graphql-hooks`
  - `cache-and-network`: The refetch function provides this behaviour it will set loading: true, but the old data will be still set until the fetch resolves.
  - `network-only` ➡️ `useQuery(QUERY, { skipCache: true })`
  - `cache-only`: Not supported
  - `no-cache` ➡️ `useQuery(QUERY, { useCache: false })`

**Not yet supported**

- `errorPolicy`: Any error will set the `error` to be truthy. See [useQuery](#useQuery) for more details.
- `pollInterval`
- `notifyOnNetworkStatusChange`
- `skip`
- `onCompleted`: Similar ability if using `useManualQery`
- `onError`: Similar ability if using `useManualQery`
- `partialRefetch`

### Query Component Render Props

```diff
- <Query query={gql`...`}>
-  {(props) => {}}
- </Query>
+ const state = useQuery(query)
```

- `props.loading` ➡️ `state.loading`

> - `props.error`: The error prop from `useQuery` is Boolean the details of the error can be found in either:

Reviewer: ```suggestion
- `props.error`: The error value from `useQuery` is Boolean, the details of the error can be found in either:
```

### Mutation Component ➡️ useMutation

```diff
- import { Mutation } from 'react-apollo'
- import gql from 'graphql-tag'
+ import { useMutation } from 'graphql-hooks'

function MyComponent() {
+ const [mutateFn, { loading, error, data }] = useMutation('...')

-  return (
-    <Mutation mutation={gql`...`}>
-     {(mutateFn, { loading, error }) => {
        if (error) return 'Error :('

        return <button disabled={loading} onClick={() => mutateFn()}>Submit</button>
-      }}
-    </Mutation>
-  )
}
```

### Mutation Props

- `mutation` ➡️ `useMutation(mutation)` - no need to wrap it in `gql`
- `variables` ➡️️ `useMutation(mutation, { variables })` or `mutateFn({ variables })`
- `ignoreResults` ➡️️️️ `const [mutateFn] = useMutation(mutation)`
- `onCompleted` ➡️ ️`mutateFn().then(onCompleted)`
- `onError` ➡️ `mutateFn().then(({ error }) => {...})`

**Not yet supported**

- `update`: Coming soon [#52](https://github.com/nearform/graphql-hooks/issues/52)
- `optimisticResponse`
- `refetchQueries`
- `awaitRefetchQueries`
- `context`

## Mutation Component Render Props

- `data` ➡️ `const [mutateFn, { data }] = useMutation`
- `loading` ➡️ `const [mutateFn, { loading }] = useMutation`

> - `error` ➡️ `const [mutateFn, { error }] = useMutation`

Reviewer: Do we need a similar description to `useQuery` here? Our `error` is a boolean

-  cache: new InMemoryCache()
- })
+ const client = new GraphQLClient({
+   url: '/graphql',
+   cache: memCache()
+ })
```

A lot of the options you'd pass to `ApolloClient` are the same as `GraphQLClient`:

- `uri` ➡️ `url`
- `fetchOptions`
- `onError` - the function signature is slightly different
- `headers`
- `fetch`
- `cache`

### ApolloProvider ➡️ ClientContext.Provider

```diff
- import { ApolloProvider } from 'react-apollo'
+ import { ClientContext } from 'graphql-hooks'

function App({ client }) {
  return (
-    <ApolloProvider client={client}>
+    <ClientContext.Provider value={client}>
       {/* children */}
+    </ClientContext.Provider>
-    </ApolloProvider>
  )
}
```

### Query Component ➡️ useQuery

```diff
- import { Query } from 'react-apollo'
- import gql from 'graphql-tag'
+ import { useQuery } from 'graphql-hooks'

function MyComponent() {
+ const { loading, error, data } = useQuery('...')

-  return (
-    <Query query={gql`...`}>
-     {({ loading, error, data}) => {
        if (loading) return 'Loading...'
        if (error) return 'Error :('

        return <div>{data}</div>
-      }}
-    </Query>
-  )
}
```

### Query Component Props

> A lot of options can be carried over as-is, or have direct replacement:

Reviewer: ```suggestion
A lot of options can be carried over as-is, or have direct replacements:
```

- `query` ➡️ `useQuery(query)`: No need to wrap the query in `gql`
- `variables` ➡️ `useQuery(query, { variables })`
- `ssr` ➡️ `useQuery(query, { ssr })`
- **Fetch Policies**: See [#75](https://github.com/nearform/graphql-hooks/issues/75) for more info
  - `cache-first`: This the default behaviour of `graphql-hooks`
  - `cache-and-network`: The refetch function provides this behaviour it will set loading: true, but the old data will be still set until the fetch resolves.
  - `network-only` ➡️ `useQuery(QUERY, { skipCache: true })`
  - `cache-only`: Not supported
  - `no-cache` ➡️ `useQuery(QUERY, { useCache: false })`

**Not yet supported**

- `errorPolicy`: Any error will set the `error` to be truthy. See [useQuery](#useQuery) for more details.
- `pollInterval`
- `notifyOnNetworkStatusChange`
- `skip`

> - `onCompleted`: Similar ability if using `useManualQery`

Reviewer 2: ```suggestion
- `onCompleted`: Similar ability if using `useManualQuery`
```

> - `onError`: Similar ability if using `useManualQery`

Reviewer 2: ```suggestion
- `onError`: Similar ability if using `useManualQuery`
```
